reimplement merge to drop mergo dependency

liggitt opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Reimplements the narrow slice of merge capability we need from mergo to allow dropping a dependency.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/107499

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/area code-organization
/sig architecture cli api-machinery
/cc @deads2k @dims

liggitt committed:
Isolate mergo use, add test coverage and error checking

liggitt committed:
Narrow merge interface to merging the same types

liggitt committed:
Implement limited merge function

liggitt committed:
Update vendor

liggitt commented:
/skip

dims commented on staging/src/k8s.io/client-go/tools/clientcmd/merge.go:
```diff
@@ -0,0 +1,80 @@
+// - non-pointer struct fields are recursively merged
+// - maps are shallow merged with src keys taking priority over dst
+// - non-zero src fields encountered during recursion that are not maps or structs overwrite and recursion stops
+func merge[T any](dst, src *T) error {
```
looks like this is equivalent to the `mergo.Merge` with `WithOverride` option

liggitt replied:
yes, that was our only use

dims commented:
@liggitt i was trying some of the test cases in the original mergo repo with the specific tag and we seem to trip up this one. can you please peek?

https://github.com/darccio/mergo/blame/v0.3.6/issue64_test.go#L19

```
{Student{"Jack", []string{}}, Student{"Tom", []string{"1"}}, []string{"1"}}
```

dims commented:
@liggitt also i don't know if we end up merging timestamps in our use case... `TestIssue38MergeWithOverwrite` and `TestOverwriteZeroSrcTime` are wonky as well in addition to the one above.

dims commented:
@liggitt can we do this for this release?

liggitt commented:
Rebased this, pulled in a tmp commit that runs all the mergo unit tests that exercise WithOverwrite invocations (including TestIssue38MergeWithOverwrite, TestOverwriteZeroSrcTime, and TestIssue64MergeSliceWithOverride, as noted above).

Reworked the implementation to make all relevant mergo unit tests pass as well.

Once this has review, I'll drop the TMP commit with those unit tests.

liggitt commented:
/hold for review and drop of TMP commit with mergo unit tests

liggitt commented:
/retest pull-kubernetes-e2e-gce

liggitt commented:
(lint jobs are complaining about the mergo unit test files that will be dropped)

liggitt commented:
/test pull-kubernetes-e2e-gce

liggitt commented:
/test pull-kubernetes-unit

dims commented:
Looks good to me @liggitt please drop the `TMP` commit and i'll add the `/lgtm`. thanks for testing with the mergo tests!

liggitt commented:
For posterity, clean unit test run at https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/126764/pull-kubernetes-unit/1839831403240361984

liggitt commented:
done

liggitt commented:
/hold cancel

dims commented:
some linter and verify-vendor errors ... looks legit

liggitt commented:
```
elseif: can replace 'else {if cond {}}' with 'else if cond {}' (gocritic)
		} else {
```

🙄 the linter is *quite* irritating at times

dims commented:
/approve
/lgtm

liggitt commented:
/retest

seans3 commented:
/triage accepted
